src/IntercomClient.php

```diff
@@ -190,108 +207,164 @@ public function put($endpoint, $json)
-            [
-            'json' => $json,
-            'auth' => $this->getAuth(),
-            'headers' => [
-                'Accept' => 'application/json'
-            ],
-            ]
-        );
-
-        $response = $this->http_client->request('DELETE', "https://api.intercom.io/$endpoint", $guzzleRequestOptions);
+        $response = $this->sendRequest('DELETE', "https://api.intercom.io/$endpoint", $json);
         return $this->handleResponse($response);
     }
 
     /**
+     * Sends GET request to Intercom API.
+     *
      * @param string $endpoint
-     * @param array  $query
-     * @return mixed
-     * @throws \GuzzleHttp\Exception\GuzzleException
+     * @param array  $queryParams
+     * @return stdClass
      */
-    public function get($endpoint, $query)
+    public function get($endpoint, $queryParams = [])
     {
-        $guzzleRequestOptions = $this->getGuzzleRequestOptions(
-            [
-            'query' => $query,
-            'auth' => $this->getAuth(),
-            'headers' => [
-                'Accept' => 'application/json'
-            ],
-            ]
-        );
+        $uri = $this->getUriFactory()->createUri("https://api.intercom.io/$endpoint");
+        if (!empty($queryParams)) {
+            $uri = $uri->withQuery(http_build_query($queryParams));
+        }
+
+        $response = $this->sendRequest('GET', $uri);
 
-        $response = $this->http_client->request('GET', "https://api.intercom.io/$endpoint", $guzzleRequestOptions);
         return $this->handleResponse($response);
     }
 
     /**
-     * Returns next page of the result.
+     * Returns the next page of the result.
      *
-     * @param  \stdClass $pages
-     * @return mixed
-     * @throws \GuzzleHttp\Exception\GuzzleException
+     * @param  stdClass $pages
+     * @return stdClass
      */
     public function nextPage($pages)
     {
-        $guzzleRequestOptions = $this->getGuzzleRequestOptions(
+        $response = $this->sendRequest('GET', $pages->next);
+        return $this->handleResponse($response);
+    }
+
+    /**
+     * Gets the rate limit details.
+     *
+     * @return array
+     */
+    public function getRateLimitDetails()
+    {
+        return $this->rateLimitDetails;
+    }
+
+    /**
+     * @return array
+     */
+    private function getRequestHeaders()
+    {
+        return array_merge(
             [
-            'auth' => $this->getAuth(),
-            'headers' => [
-                'Accept' => 'application/json'
+                'Accept' => 'application/json',
+                'Content-Type' => 'application/json'
             ],
-            ]
+            $this->extraRequestHeaders
         );
+    }
 
-        $response = $this->http_client->request('GET', $pages->next, $guzzleRequestOptions);
-        return $this->handleResponse($response);
+    /**
+     * @return ClientInterface
+     */
+    private function getClient()
+    {
+        if (empty($this->client)) {
```

Comment:
`=== null` instead ? And IMO, you should register the client on the construct, and let it override. So that you can even remove the need for a `getClient` call.

Reply:
The reasoning behind adding the `getClient` method was to avoid doing the discovery if it's not needed, but this might be an overkill. I will move these to the constructor 😃

Reply:
Maybe you could add a nullable client in the constructor, so that you try to discover if Bill is given ? :)

Reply:
That's a good point! However, I'd prefer to keep the constructor logic as little as possible, it's already pretty big. I'd also prefer to keep the `setClient` method, as that would reduce the upgrade friction (that method is there in the current version of the SDK). What do you think?

Reply:
Adding the constructor parameter does not prevent you from adding the `serClient` but then I don't really see the problem with defaulting to Discovery and having to call `setClient` to change that

Reply:
Agree, let's keep it simple 👍I'll let it do the discovery in the constructor and then we'll allow to override